exla/lib/exla/defn.ex

```diff
@@ -2629,6 +2649,49 @@ defmodule EXLA.Defn do
     end
   end
 
+  defp recur_shared_ids(
+         expr,
+         other_ids,
+         %{scope_ids: ids} = state,
+         cache
+       ) do
+    {_, ids_args} =
+      Composite.reduce(expr, {%{}, %{}}, fn node, acc ->
+        {_, acc} = collect_args(node, acc, {ids, other_ids})
+        acc
+      end)
+
+    # sorted_ids_args = Enum.sort_by(ids_args, fn {_id, {i, _old, _new}} -> i end)
+
+    Enum.reduce(ids_args, cache, fn {_, {_, old, _}}, cache ->
+      {_, cache} = recur_operator(old, state, cache)
+      cache
+    end)
+  end
+
+  defp to_mlir_if_branch(
+         bool,
+         node,
+         expr,
+         current_ids,
+         state,
+         cache
+       ) do
+    comp_state = %{state | scope_ids: current_ids}
+
+    Value.set_if_block(node, bool)
+    {res, cache} = recur_composite(expr, &cast_pred_to_u8/1, comp_state, cache)
+
+    if token = get_token(cache) do
+      Value.variadic_return([token, res], true)
+    else
+      Value.variadic_return([res], true)
+    end
+
+    Function.reset_region(state.builder)
+    cache
```

Comment:
If we indeed don't need to return the cache, we can simply do:

```suggestion
    {res, _cache} = recur_composite(expr, &cast_pred_to_u8/1, comp_state, cache)

    if token = get_token(cache) do
      Value.variadic_return([token, res], true)
    else
      Value.variadic_return([res], true)
    end

    Function.reset_region(state.builder)
```